### PR TITLE
Add PushDownAggregate optimization

### LINKED ISSUE #4772: `PushDownProjection` doesn't push down the empty projection to the child plans of `Union`
BTW, I'm not sure if it was a bug, but the `PushDownProjection` doesn't push down the empty projection to the child plans of `Union` in this example. Maybe @jackwener knows more about it.

_Originally posted by @HaoYang670 in https://github.com/apache/arrow-datafusion/pull/4753#discussion_r1058696036_

### PR DESCRIPTION
**Pull request by mslapek**
# Which issue does this PR close?

Closes #4772

# Rationale for this change

See the issue.

# What changes are included in this PR?

Added a new optimization `PushDownAggregate`

Added it to the default list in `optimizer.rs`.

# Are these changes tested?

Unit tests inside  `push_down_aggregate.rs`.

Added tests to `pg_compat_union.slt` to ensure, that computation results are still correct.

# Are there any user-facing changes?

No.

### CODE REVIEW

**Comment by mingmwang** [comment]
No, I'm afraid this optimization is useless. Could you please check the physical plan ?
I think in the physical plan phase, there was already the partial agg and final agg and actually the partial agg is already pushed down.

**Comment by mslapek** [comment]
@mingmwang Thank you for your observation. Let's make an experiment...

## Experiment

Copied `arrow-datafusion/datafusion/core/tests/data/aggregate_simple.csv` to `aggregate_simple.csv`.

Script:

```rust
use datafusion::prelude::*;

#[tokio::main]
async fn main() -> datafusion::error::Result<()> {
    let ctx = SessionContext::new();

    ctx.register_csv(
        "my_table",
        "datafusion/core/tests/data/aggregate_simple.csv",
        CsvReadOptions::default(),
    ).await?;
    ctx.register_csv(
        "my_table2",
        "datafusion/core/tests/data/aggregate_simple2.csv",
        CsvReadOptions::default(),
    ).await?;

    let sql = "SELECT COUNT(*) FROM (SELECT * from my_table UNION ALL SELECT * from my_table2)";
    let df = ctx.sql(sql).await?;
    df.clone().explain(true, false)?.show().await?;
    df.show().await?;

    Ok(())
}

```

Optimized logical and physical plans are below.

## Without `push_down_aggregate`

*Explain after optimizations*

```
...
| logical_plan                                               | Aggregate: groupBy=[[]], aggr=[[COUNT(UInt8(1))]]                                                                                                                        |
|                                                            |   Union                                                                                                                                                                  |
|                                                            |     TableScan: my_table projection=[c1]                                                                                                                                  |
|                                                            |     Projection: my_table2.c1 AS my_table.c1                                                                                                                              |
|                                                            |       TableScan: my_table2 projection=[c1]
...
| physical_plan                                              | AggregateExec: mode=Final, gby=[], aggr=[COUNT(UInt8(1))]                                                                                                                |
|                                                            |   CoalescePartitionsExec                                                                                                                                                 |
|                                                            |     AggregateExec: mode=Partial, gby=[], aggr=[COUNT(UInt8(1))]                                                                                                          |
|                                                            |       RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=2                                                                                              |
|                                                            |         UnionExec                                                                                                                                                        |
|                                                            |           CsvExec: files={1 group: [[/home/hello/repo/arrow-datafusion/datafusion/core/tests/data/aggregate_simple.csv]]}, has_header=true, limit=None, projection=[c1]    |
|                                                            |           ProjectionExec: expr=[c1@0 as my_table.c1]                                                                                                                     |
|                                                            |             CsvExec: files={1 group: [[/home/hello/repo/arrow-datafusion/datafusion/core/tests/data/aggregate_simple2.csv]]}, has_header=true, limit=None, projection=[c1] |
|                                                            |                                                                                                                                                                          |
+------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

These CSV files will be naively mixed together with `RoundRobinBatch` - only to count the rows at the end! 😧

> actually the partial agg is already pushed down.

❗️ No, it isn't...

## With `push_down_aggregate`

*Explain after optimizations*

```
...
| logical_plan                                               | Projection: SUM(COUNT(aggr_1)) AS COUNT(UInt8(1))                                                                                                                                |
|                                                            |   Aggregate: groupBy=[[]], aggr=[[SUM(COUNT(aggr_1))]]                                                                                                                           |
|                                                            |     Union                                                                                                                                                                        |
|                                                            |       Aggregate: groupBy=[[]], aggr=[[COUNT(aggr_1)]]                                                                                                                            |
|                                                            |         Projection: UInt8(1) AS aggr_1                                                                                                                                           |
|                                                            |           TableScan: my_table projection=[c1]                                                                                                                                    |
|                                                            |       Aggregate: groupBy=[[]], aggr=[[COUNT(aggr_1)]]                                                                                                                            |
|                                                            |         Projection: UInt8(1) AS aggr_1                                                                                                                                           |
|                                                            |           TableScan: my_table2 projection=[c1]
...
| physical_plan                                              | ProjectionExec: expr=[SUM(COUNT(aggr_1))@0 as COUNT(UInt8(1))]                                                                                                                   |
|                                                            |   AggregateExec: mode=Final, gby=[], aggr=[SUM(COUNT(aggr_1))]                                                                                                                   |
|                                                            |     CoalescePartitionsExec                                                                                                                                                       |
|                                                            |       AggregateExec: mode=Partial, gby=[], aggr=[SUM(COUNT(aggr_1))]                                                                                                             |
|                                                            |         UnionExec                                                                                                                                                                |
|                                                            |           AggregateExec: mode=Final, gby=[], aggr=[COUNT(aggr_1)]                                                                                                                |
|                                                            |             CoalescePartitionsExec                                                                                                                                               |
|                                                            |               AggregateExec: mode=Partial, gby=[], aggr=[COUNT(aggr_1)]                                                                                                          |
|                                                            |                 ProjectionExec: expr=[1 as aggr_1]                                                                                                                               |
|                                                            |                   RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=1                                                                                          |
|                                                            |                     CsvExec: files={1 group: [[/home/hello/repo/arrow-datafusion/datafusion/core/tests/data/aggregate_simple.csv]]}, has_header=true, limit=None, projection=[c1]  |
|                                                            |           AggregateExec: mode=Final, gby=[], aggr=[COUNT(aggr_1)]                                                                                                                |
|                                                            |             CoalescePartitionsExec                                                                                                                                               |
|                                                            |               AggregateExec: mode=Partial, gby=[], aggr=[COUNT(aggr_1)]                                                                                                          |
|                                                            |                 ProjectionExec: expr=[1 as aggr_1]                                                                                                                               |
|                                                            |                   RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=1                                                                                          |
|                                                            |                     CsvExec: files={1 group: [[/home/hello/repo/arrow-datafusion/datafusion/core/tests/data/aggregate_simple2.csv]]}, has_header=true, limit=None, projection=[c1] |
|
```

There is no mixing of the file contents.
To `AggregateExec: mode=Partial, gby=[], aggr=[SUM(COUNT(aggr_1))]` are going only two rows.

**Comment by mslapek** [comment]
@mingmwang You gave a great idea - it should be **a physical plan optimization!** 💡

The optimization should push down partial aggregate though union:

```
...
| logical_plan                                               | Aggregate: groupBy=[[]], aggr=[[COUNT(UInt8(1))]]                                                                                                                        |
|                                                            |   Union                                                                                                                                                                  |
|                                                            |     TableScan: my_table projection=[c1]                                                                                                                                  |
|                                                            |     Projection: my_table2.c1 AS my_table.c1                                                                                                                              |
|                                                            |       TableScan: my_table2 projection=[c1]
...
| physical_plan                                              | AggregateExec: mode=Final, gby=[], aggr=[COUNT(UInt8(1))]                                                                                                                |
|                                                            |   CoalescePartitionsExec                                                                                                                                                 |                                                                                                        |
|                                                            |       RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=2                                                                                              |
|                                                            |         UnionExec      
|                                                            |           AggregateExec: mode=Partial, gby=[], aggr=[COUNT(UInt8(1))]                                                                                                                                                    |
|                                                            |             CsvExec: files={1 group: [[/home/hello/repo/arrow-datafusion/datafusion/core/tests/data/aggregate_simple.csv]]}, has_header=true, limit=None, projection=[c1]    |
|                                                            |           AggregateExec: mode=Partial, gby=[], aggr=[COUNT(UInt8(1))]   
|                                                            |             ProjectionExec: expr=[c1@0 as my_table.c1]                                                                                                                     |
|                                                            |               CsvExec: files={1 group: [[/home/hello/repo/arrow-datafusion/datafusion/core/tests/data/aggregate_simple2.csv]]}, has_header=true, limit=None, projection=[c1] |
|                                                            |                                                                                                                                                                          |
+------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

Optimization in the logical plan gave unnecessary doubled final-partial aggregate pair.

@mingmwang What do you think?

**Comment by mingmwang** [comment]
> @mingmwang You gave a great idea - it should be **a physical plan optimization!** 💡
> 
> The optimization should push down partial aggregate though union:
> 
> ```
> ...
> | logical_plan                                               | Aggregate: groupBy=[[]], aggr=[[COUNT(UInt8(1))]]                                                                                                                        |
> |                                                            |   Union                                                                                                                                                                  |
> |                                                            |     TableScan: my_table projection=[c1]                                                                                                                                  |
> |                                                            |     Projection: my_table2.c1 AS my_table.c1                                                                                                                              |
> |                                                            |       TableScan: my_table2 projection=[c1]
> ...
> | physical_plan                                              | AggregateExec: mode=Final, gby=[], aggr=[COUNT(UInt8(1))]                                                                                                                |
> |                                                            |   CoalescePartitionsExec                                                                                                                                                 |                                                                                                        |
> |                                                            |       RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=2                                                                                              |
> |                                                            |         UnionExec      
> |                                                            |           AggregateExec: mode=Partial, gby=[], aggr=[COUNT(UInt8(1))]                                                                                                                                                    |
> |                                                            |             CsvExec: files={1 group: [[/home/hello/repo/arrow-datafusion/datafusion/core/tests/data/aggregate_simple.csv]]}, has_header=true, limit=None, projection=[c1]    |
> |                                                            |           AggregateExec: mode=Partial, gby=[], aggr=[COUNT(UInt8(1))]   
> |                                                            |             ProjectionExec: expr=[c1@0 as my_table.c1]                                                                                                                     |
> |                                                            |               CsvExec: files={1 group: [[/home/hello/repo/arrow-datafusion/datafusion/core/tests/data/aggregate_simple2.csv]]}, has_header=true, limit=None, projection=[c1] |
> |                                                            |                                                                                                                                                                          |
> +------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> ```
> 
> Optimization in the logical plan gave unnecessary doubled final-partial aggregate pair.
> 
> @mingmwang What do you think?

I think even during the physical planning phase, push down the partial aggregate though repartition + union might not give performance benefit, because it might reduce the parallism of the partial aggregate.

**Comment by mslapek** [comment]
My intuitions says, that 2 partial aggregates in **PARALLEL** branches would be more susceptible to parallelization - than one big partial aggregate. 🧐

Could you clarify, why that's not the case?

**Comment by mingmwang** [comment]
> My intuitions says, that 2 partial aggregates in **PARALLEL** branches would be more susceptible to parallelization - than one big partial aggregate. 🧐
> 
> Could you clarify, why that's not the case?

This is related to how UnionExec is actually executed:
For example, if we have a plan fragment like below, if each CsvExec's parallelization is 1, then the AggregateExec's parallelization is 2

```
AggregateExec(2)
    UnionExec(2)
       CsvExec(1)
       CsvExec(1)
```

If we push down the  AggregateExec through UnionExec like below,  the total parallelization of AggregateExec is still 2.
This is what I mean the pushdown is useless.

```
 UnionExec(2)
    AggregateExec(1)
       CsvExec(1)
   AggregateExec(1)
       CsvExec(1)
```

Now we take the RepartitionExec into consideration, if the RepartitionExec change the parallelization to 4, the parallelization of AggregateExec is 4


```
AggregateExec(4)
   RepartitionExec(4)
       UnionExec(2)
         CsvExec(1)
         CsvExec(1)
```
If we push down the AggregateExec  through RepartitionExec + UnionExec ,  it will reduce the parallelization of AggregateExec to 2 and will not get performance benefits

```
 RepartitionExec(4)
     UnionExec(2)
        AggregateExec(1)
           CsvExec(1)
        AggregateExec(1)
           CsvExec(1)
```

Hope this explains.

**Review by alamb — COMMENTED** [review]
@Jefffrey @HaoYang670  @jackwener  I wonder if you have time to review this PR (as you filed / commented on the original ticket https://github.com/apache/arrow-datafusion/issues/4772)

**Comment by mslapek** [comment]
@mingmwang Thank you for a thorough explanation! 🙂 It makes a perfect sense.

Yeah. Union is only a concatenated list of partitions... itself it does no performance impact - that's what I've forgotten! 😬

So the only performance issue is: when `RepartitionExec(4)` can be avoided? When we have in all branches equally-sized partitions, and their amount allows for a sensible parallelization - then the reshuffle could be avoided.

But:
1. it is not related in a specific way to the union, it's a general issue
2. it would require data from some Metastore

---

To sum up, it looks like this PR should be closed without merge. 😭

**Comment by mingmwang** [comment]
I think you can close this PR. 
But you can file another ticket to track the `RepartitionExec` enhancement.  
It is better if you can give some real SQL examples with the physical plan fragments generated by the DataFusion optimizer which can prove the `RepartitionExec` is unnecessary.

**Comment by alamb** [comment]
Closing per comments above
